src/types/pick.ts

```diff
@@ -0,0 +1,103 @@
+import { BaseObjectLikeTypeImpl, createType } from '../base-type.js';
+import type {
+    BasicType,
+    MessageDetails,
+    OneOrMore,
+    PossibleDiscriminator,
+    Properties,
+    PropertiesInfo,
+    Result,
+    Type,
+    TypeImpl,
+    TypeOf,
+    ValidationOptions,
+    Visitor,
+} from '../interfaces.js';
+import { decodeOptionalName } from '../utils/collection-utils.js';
+import { define } from '../utils/define.js';
+import { unknownRecord } from './unknown.js';
+import { hasOwnProperty } from '../utils/type-utils.js';
+import { prependPathToDetails } from '../utils/failure-utils.js';
+import { interfaceStringify } from '../utils/stringifiers.js';
+import { propsInfoToProps } from './union.js';
+
+export class PickType<Type extends BaseObjectLikeTypeImpl<unknown>, ResultType> extends BaseObjectLikeTypeImpl<ResultType> {
+    readonly name: string;
+    readonly isDefaultName: boolean;
+    readonly props: Properties;
+    override propsInfo: PropertiesInfo;
+    constructor(
+        readonly type: Type,
+        readonly keys: OneOrMore<keyof TypeOf<Type>>,
+        name?: string,
+    ) {
+        super();
+        this.isDefaultName = !name;
+        this.name = name || `Pick<${type.name}, ${keys.map(k => `'${k.toString()}'`).join(' | ')}>`;
+        const stringKeys = this.keys.map(k => k.toString()) as [string, ...string[]];
+        this.propsInfo = pickPropertiesInfo(this.type.propsInfo, stringKeys);
+        this.props = propsInfoToProps(this.propsInfo);
+        this.possibleDiscriminators = this.type.possibleDiscriminators;
+    }
+    override possibleDiscriminators: readonly PossibleDiscriminator[] = this.type.possibleDiscriminators; // TODO: Filter
+    override basicType!: BasicType;
+    override typeConfig: unknown;
+
+    protected override typeValidator(input: unknown, options: ValidationOptions): Result<ResultType> {
+        if (!unknownRecord.is(input)) {
+            return this.createResult(input, undefined, { kind: 'invalid basic type', expected: 'object' });
+        }
+
+        const constructResult: Record<string, unknown> = {};
+        const details: MessageDetails[] = [];
+        for (const [key, innerType] of this.propsArray) {
+            const missingKey = !hasOwnProperty(input, key);
+            const partialKey = this.propsInfo[key]?.partial;
+
+            if (missingKey) {
+                partialKey || details.push({ kind: 'missing property', property: key, type: innerType });
+                continue;
+            }
+
+            const innerResult = innerType.validate(input[key], options);
+            if (innerResult.ok) {
+                constructResult[key] = innerResult.value;
+            } else {
+                details.push(...prependPathToDetails(innerResult, key));
+            }
+        }
+        return this.createResult(input, options.mode === 'construct' ? constructResult : input, details);
+    }
```

Comment:
This method feels kind of familiar, doesn't it? This is currently great to prove the concept. One of the things you had to change here compared to the implementation in `InterfaceType` is that "partial" is now an aspect of a single property, instead of the entire type.

I think we should consider refactoring `InterfaceType` to support that as well. All the ingredients are already in place. This could have the following benefits: (I haven't tried it though)
- No more need for a seperate `PickType`, we can now delegate entirely to `InterfaceType` (sometimes combined with a `UnionType` when the input type is a union) from within the `pick` function
- `withOptional` then returns an `InterfaceType` instead of an `IntersectionType` makes sense I guess?[^1]
- we can support config options like `strictMissingKeys` etc

If we decide to do that, we should do so in another branch. Let's discuss that offline.

[^1]: Currently, objects with both mandatory and optional properties are implemented as an intersection of one object with the mandatory properties and one object with the optional properties.

Reply:
Nieuwe issue voor aangemaakt #93